### PR TITLE
feat(retrieval): hybrid BM25+cosine retrieve helper with RRF fusion (G0.4-T4)

### LINKED ISSUE #261: Task: G0.4-T4 retrieve helper — hybrid BM25 (Postgres FTS) + cosine (pgvector) + RRF fusion (k=60)
Parent initiative: #225

## Target

`retrieve` helper — hybrid BM25 + cosine RRF retrieval:

- `backend/src/meho_backplane/retrieval/retriever.py` — new module with `retrieve` function + `RetrievalHit` Pydantic model.
- `backend/tests/test_retrieval_retriever.py` — new file. Unit + integration tests.

## Context

T4 ships the **shared read path**. G4 `meho kb search`, G5 `meho recall`, future agent-grounding flows all call `retrieve` and consume `RetrievalHit` results. The retriever's correctness is load-bearing.

**Hybrid retrieval per v0.1-spec L391.** Two retrieval signals, fused via Reciprocal Rank Fusion:

1. **BM25** via Postgres `to_tsvector` + `ts_rank_cd` against the GIN index (T1 created).
2. **Cosine** via pgvector's `<=>` operator against the IVFFlat index (T1 created).
3. **RRF fusion** with k=60 (Microsoft paper default).

**Pull top 50 from each, fuse to top `limit`.** Per the RRF paper: rank the result lists, sum `1/(k + rank)` per document across lists, sort. Score is the fused value; per-signal scores carried for observability.

**Query plan sketch:**

```sql
-- BM25 candidates (top 50)
WITH bm25_candidates AS (
  SELECT id, ts_rank_cd(to_tsvector('english', body), plainto_tsquery('english', :query)) AS bm25_score
  FROM documents
  WHERE tenant_id = :tenant_id
    AND (:source IS NULL OR source = :source)
    AND (:kind IS NULL OR kind = :kind)
    AND to_tsvector('english', body) @@ plainto_tsquery('english', :query)
  ORDER BY bm25_score DESC
  LIMIT 50
),
-- Cosine candidates (top 50)
cosine_candidates AS (
  SELECT id, 1 - (embedding <=> :query_embedding) AS cosine_score
  FROM documents
  WHERE tenant_id = :tenant_id
    AND (:source IS NULL OR source = :source)
    AND (:kind IS NULL OR kind = :kind)
  ORDER BY embedding <=> :query_embedding
  LIMIT 50
)
SELECT ... -- union, dedupe, RRF rank, top :limit
```

In Python, we run the two queries separately, fuse in-process, then fetch the full Document rows for the top-k by ID. Avoids over-fetching full document bodies for non-final candidates.

**Tenant scoping.** Caller passes `tenant_id`; every query filters by it.

**Source / kind filters.** Optional. G4's `meho kb search` calls `retrieve(source="kb")`; G5's `meho recall` calls `retrieve(source="memory", kind="memory-user")` to scope. Future global queries pass None.

**Empty corpus behavior.** Both signals return empty → `retrieve` returns `[]`. Document the no-results contract clearly in the function docstring.

**Scoring semantics.**
- `bm25_score`: 0.0 to ~1.0 (Postgres `ts_rank_cd` is unbounded; we clamp via normalization step in the query). Score is recall-shaped.
- `cosine_score`: 0.0 to 1.0 (we use `1 - distance` since `<=>` returns cosine *distance*).
- `fused_score`: RRF score, typically 0.0 to ~0.05 (small numbers; k=60 makes the magnitude tight but ordering meaningful). Callers should rank, not threshold.

## Current state

After T1 + T2 + T3:
- `documents` table populated with text + embeddings.
- No retrieval surface; callers would have to write hybrid queries themselves.

## Desired state

`backend/src/meho_backplane/retrieval/retriever.py`:

```python
# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2026 evoila Group

from typing import Any
from uuid import UUID

import structlog
from pydantic import BaseModel, ConfigDict
from sqlalchemy import select, text
from sqlalchemy.ext.asyncio import AsyncSession

from meho_backplane.db.engine import async_session
from meho_backplane.db.models import Document
from meho_backplane.retrieval.embedding import get_embedding_service

logger = structlog.get_logger()


class RetrievalHit(BaseModel):
    model_config = ConfigDict(frozen=True)

    document_id: UUID
    source: str
    source_id: str
    kind: str
    body: str
    metadata: dict[str, Any]
    fused_score: float
    bm25_score: float | None  # None if doc didn't appear in BM25 top-50
    cosine_score: float | None  # None if doc didn't appear in cosine top-50
    bm25_rank: int | None
    cosine_rank: int | None


_RRF_K = 60  # Microsoft RRF paper default


async def retrieve(
    tenant_id: UUID,
    query: str,
    source: str | None = None,
    kind: str | None = None,
    limit: int = 10,
    session: AsyncSession | None = None,
) -> list[RetrievalHit]:
    """Hybrid BM25 + cosine retrieval with RRF fusion. Tenant-scoped.

    Returns top `limit` hits by RRF score. Empty list if no matches.
    """
    if session is not None:
        return await _retrieve_in_session(session, tenant_id, query, source, kind, limit)
    async with async_session() as s:
        return await _retrieve_in_session(s, tenant_id, query, source, kind, limit)


async def _retrieve_in_session(
    session: AsyncSession,
    tenant_id: UUID,
    query: str,
    source: str | None,
    kind: str | None,
    limit: int,
) -> list[RetrievalHit]:
    query_embedding = await get_embedding_service().encode_one(query)

    # BM25 top 50
    bm25_sql = text("""
        SELECT id, ts_rank_cd(to_tsvector('english', body), plainto_tsquery('english', :query)) AS score
        FROM documents
        WHERE tenant_id = :tenant_id
          AND (:source IS NULL OR source = :source)
          AND (:kind IS NULL OR kind = :kind)
          AND to_tsvector('english', body) @@ plainto_tsquery('english', :query)
        ORDER BY score DESC
        LIMIT 50
    """)
    bm25_rows = (await session.execute(bm25_sql, {
        "query": query, "tenant_id": str(tenant_id), "source": source, "kind": kind,
    })).fetchall()

    # Cosine top 50 (always returns 50 if corpus has 50+ docs in scope)
    cosine_sql = text("""
        SELECT id, 1 - (embedding <=> CAST(:emb AS vector)) AS score
        FROM documents
        WHERE tenant_id = :tenant_id
          AND (:source IS NULL OR source = :source)
          AND (:kind IS NULL OR kind = :kind)
        ORDER BY embedding <=> CAST(:emb AS vector)
        LIMIT 50
    """)
    cosine_rows = (await session.execute(cosine_sql, {
        "emb": query_embedding, "tenant_id": str(tenant_id), "source": source, "kind": kind,
    })).fetchall()

    # RRF fusion
    bm25_ranks = {row.id: (i + 1, row.score) for i, row in enumerate(bm25_rows)}
    cosine_ranks = {row.id: (i + 1, row.score) for i, row in enumerate(cosine_rows)}
    candidate_ids = set(bm25_ranks) | set(cosine_ranks)
    scored = []
    for cid in candidate_ids:
        b = bm25_ranks.get(cid)
        c = cosine_ranks.get(cid)
        fused = 0.0
        if b: fused += 1.0 / (_RRF_K + b[0])
        if c: fused += 1.0 / (_RRF_K + c[0])
        scored.append((cid, fused, b, c))
    scored.sort(key=lambda x: x[1], reverse=True)
    top = scored[:limit]

    if not top:
        return []

    # Fetch full documents
    ids = [t[0] for t in top]
    doc_result = await session.execute(select(Document).where(Document.id.in_(ids)))
    docs_by_id = {d.id: d for d in doc_result.scalars().all()}

    hits = []
    for cid, fused, b, c in top:
        doc = docs_by_id[cid]
        hits.append(RetrievalHit(
            document_id=doc.id,
            source=doc.source,
            source_id=doc.source_id,
            kind=doc.kind,
            body=doc.body,
            metadata=doc.metadata,
            fused_score=fused,
            bm25_score=b[1] if b else None,
            cosine_score=c[1] if c else None,
            bm25_rank=b[0] if b else None,
            cosine_rank=c[0] if c else None,
        ))
    return hits
```

## Acceptance criteria

- [ ] `retrieve(tenant_id, "kubernetes")` against an indexed corpus returns ordered hits.
- [ ] RRF fusion verified: index 5 docs where doc-A is BM25's #1 but cosine's #20, doc-B is cosine's #1 but BM25's #20; assert RRF top is *neither* pure-BM25-top-1 nor pure-cosine-top-1 but a fused result.
- [ ] Tenant scoping verified: docs in tenant-A are invisible to a `retrieve(tenant_b, ...)` call.
- [ ] `source` filter works (e.g. `retrieve(..., source="kb")` returns only kb-source docs).
- [ ] `kind` filter works.
- [ ] Empty corpus → returns `[]`, not an error.
- [ ] Query with no semantic or lexical match → returns `[]`.
- [ ] Limit parameter respected.
- [ ] Each hit has `fused_score > 0`; bm25_score and cosine_score may be None if doc only appeared in one signal's top-50.
- [ ] Latency: 100-doc corpus, 5 queries, p95 < 100ms (informational, not enforced SLO).
- [ ] CI green: ruff + mypy + pytest.

## Out of scope

- Reranking (BAAI/bge-reranker, ColBERT) — v0.2 ships RRF only.
- Filtering by metadata JSONB fields — v0.2 supports source/kind only.
- Cross-tenant retrieval / global search — explicitly disallowed.
- Streaming retrieval — single batch response.
- Query expansion / synonym handling.
- Per-query embedding caching — every query re-embeds; v0.2.next can add an LRU.

## References

- Parent Initiative: [#225](https://github.com/evoila/meho/issues/225).
- Depends on: T1 (table + indexes), T2 (embedding service), T3 (indexer — needed for test fixtures).
- v0.1-spec §"Knowledge layer" L382-401.
- RRF paper: https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf
- pgvector cosine distance operator `<=>`: https://github.com/pgvector/pgvector#vector-functions
- Postgres FTS docs: https://www.postgresql.org/docs/16/textsearch.html
- AI engineering best-practices: `.claude/skills/implement-issue/ai_engineering_best_practices.md` — RRF design, retrieval evaluation discipline.
- Python best-practices: `.claude/skills/explore-and-refactor/python_best_practices.md` — async SQLAlchemy raw SQL, Pydantic result models.

### PR DESCRIPTION
**Pull request by ikaric**
## Summary

Lands **T4 of Initiative #225** (G0.4 Retrieval substrate) — the shared read path G4 (`meho kb search`), G5 (`meho recall`), and future agent-grounding flows all consume. Correctness here is load-bearing for every retrieval-quality assessment downstream.

Closes #261. Rebased onto current `main` after #298 (T3) merged.

## Algorithm (per v0.1-spec L391)

Two retrieval signals merged by Reciprocal Rank Fusion (Microsoft 2009, `k=60`):

1. **BM25** via Postgres FTS: `ts_rank_cd(to_tsvector('english', body), plainto_tsquery('english', :query))` filtered by `@@`, top 50 by score desc. Uses the `documents_body_fts_idx` GIN index from migration `0003`.
2. **Cosine** via pgvector: `1 - (embedding <=> CAST(:emb AS vector))` top 50 by distance asc. Uses the `documents_embedding_idx` IVFFlat index from migration `0003`.
3. **RRF fusion** in Python: `fused_score(d) = sum over signals s where d appears: 1.0 / (RRF_K + rank(d, s))`. 1-based ranks; documents in both signals' top-50 receive two contributions, documents in only one receive one. Sort by `fused_score` desc, take top `limit`.

Each signal's candidate set is bounded to `CANDIDATE_LIMIT = 50` so the in-process fusion is O(N) over candidates, not over the corpus.

**Why RRF and not weighted-sum**: no per-signal score normalisation (ts_rank_cd is unbounded, cosine is 0-1), no per-query weight tuning, no training data. A reranker model is the v0.2.next escape hatch if corpus-recall numbers show RRF under-performs.

## Tenant scoping

Every query filters by `tenant_id` in both SQL statements. No API surface omits tenant scoping — the caller passes `tenant_id` as a required parameter, the helper threads it into both queries, and the unique composite index on `(tenant_id, source, source_id)` guarantees no cross-tenant collision in the indexed corpus.

Optional `source` (e.g. `"kb"` vs `"memory"`) and `kind` (e.g. `"kb-entry"` vs `"memory-user"`) filters narrow within a tenant via the `CAST(:source AS text) IS NULL OR source = :source` pattern.

## RetrievalHit shape

Frozen pydantic v2 model with both the fused score and per-signal scores + ranks. Callers tuning embedding-model choice want to see whether the top hit had contributions from both signals or just one. `None` signals "didn't appear in that signal's top-50".

## Tests

`backend/tests/test_retrieval_retriever.py` (new) — 11 unit tests:

- RRF fusion math (no DB, no embedding): empty inputs, BM25-only, cosine-only, intersection doc outranking singletons (the load-bearing "fusion outperforms either signal alone" claim), limit truncation, `None`-score gracefulness, and the `RRF_K = 60` / `CANDIDATE_LIMIT = 50` constants locked in.
- `_coerce_uuid` accepts UUID instances + hex strings — the driver-portability path between asyncpg (returns UUID) and SQLite (returns str).
- `RetrievalHit` is frozen (mutation raises).
- `_FusedEntry` `__slots__` + None defaults regression.

PG-real coverage (the SQL bindings actually work against pgvector + tenant boundary + filters + empty-corpus) is deferred to T6's `tests/integration/test_retrieval_e2e.py`. The Docker-gated PG tests need an async-fixture-managed event loop because the cross-loop asyncpg cleanup pattern in sync `asyncio.run` tests hits "Event loop is closed" on container teardown. T6's existing integration conftest (`pg_engine` async fixture) is the right shape.

## Test plan

- [ ] `ruff check .` — clean.
- [ ] `ruff format --check .` — clean.
- [ ] `mypy --strict src/` — clean across 42 source files.
- [ ] `pytest tests/` — 370 passed, 1 skipped (slow), 1 pre-existing macOS-only `test_observability` failure unchanged.
- [ ] `scripts/ci/check_migration_compat.py` — exit 0.
- [ ] CI green.

## Stack context

T5 (#262) and T6 (#263) are implemented locally on stacked branches. Each opens for review only after its predecessor merges, keeping at most one G0.4 PR open at a time.

## Summary by CodeRabbit

* **New Features**
  * Hybrid retrieval combining keyword (BM25) and semantic (vector) signals with unified fused ranking and tenant-scoped filtering
  * Result payloads now include fused score plus per-signal scores and ranks for observability

* **Tests**
  * Added unit tests covering fusion logic, limits/edge cases, UUID handling, and immutability of retrieval results

* **Documentation**
  * Updated backend docs describing retrieval behavior, response shape, and failure/empty-corpus semantics

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/299)

### CODE REVIEW

File: backend/tests/test_retrieval_retriever.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+
+def test_fused_entry_starts_with_none_scores() -> None:
+    """``_FusedEntry`` defaults: all score / rank fields ``None``, fused 0.0."""
+    e = _FusedEntry(uuid.uuid4())
+    assert e.bm25_score is None
+    assert e.cosine_score is None
+    assert e.bm25_rank is None
+    assert e.cosine_rank is None
+    assert e.fused_score == 0.0
```
Comment (zdamir) [inline backend/tests/test_retrieval_retriever.py]: **[🛑 B1]** — Use `pytest.approx` for the `fused_score == 0.0` assertion (Sonar S1244 / drives Quality Gate FAIL)

**What:** Line 256 asserts `e.fused_score == 0.0` exactly.

**Why it matters:** SonarCloud's Python rule `S1244` (MAJOR / BUG) flags this and the project's Quality Gate fails on "C Reliability Rating on New Code (required ≥ A)" — Sonar attributes the rating drop to this assertion. CI's `SonarCloud Code Analysis` check is `FAILURE` on the current head solely on this gate. The same `pytest.approx` pattern is already used elsewhere in this file (e.g. line 195: `assert fused[0].fused_score == pytest.approx(1.0 / (RRF_K + 1))`), so the inconsistency itself is the smell that Sonar is calibrated against. The float math here is technically deterministic (the `_FusedEntry` default is `0.0` and no arithmetic touches it before the assertion), so the bug-class is over-flagged — but the CI gate is what counts.

**Required change:**

```diff
-    assert e.fused_score == 0.0
+    assert e.fused_score == pytest.approx(0.0)
```

`pytest` is already imported in this file (line 12) and `pytest.approx` is the established pattern for every other float assertion in the suite.

**Source:** SonarCloud check on this PR — https://sonarcloud.io/dashboard?id=evoila_meho&pullRequest=299. Sonar rule `python:S1244`. Consistent pattern at lines 195, 213, 230 (etc.) in the same file.

**Review by zdamir — CHANGES_REQUESTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #261 — G0.4-T4 hybrid BM25 + cosine `retrieve` helper with Reciprocal Rank Fusion
**Parent initiative:** #225 — G0.4 Retrieval substrate
**Parent goal:** #221 — G0 Foundational substrate
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** REQUEST_CHANGES — SonarCloud Quality Gate FAIL on the float-equality assertion drives the merge gate; two CR-flagged findings already inline.
**Head SHA reviewed:** `b260196`
**AI reviewer (CodeRabbit):** received 1 inline Major (`retrieve()` limit guard) + 1 inline Critical (`_row` type signature)

---

### Acceptance criteria

| # | Criterion (from issue #261) | Status | How verified |
|---|---|---|---|
| 1 | RRF fusion math correct (Microsoft 2009, k=60, 1-based ranks, two-signal intersection sum) | ✅ | `test_rrf_k_constant_matches_paper`, `test_rrf_fuse_handles_empty_inputs`, `test_rrf_fuse_intersection_outranks_singletons`, plus 6 more |
| 2 | Per-signal candidate set bounded to `CANDIDATE_LIMIT = 50` | ✅ | Constant locked in `test_rrf_k_constant_matches_paper`; SQL `LIMIT :limit` parameterised with `CANDIDATE_LIMIT` |
| 3 | `RetrievalHit` exposes fused score + per-signal scores + ranks (None for missing) | ✅ | `test_rrf_fuse_handles_none_scores`, `test_retrieval_hit_is_frozen` |
| 4 | Tenant scoping in both SQL statements | ✅ | Both BM25 and cosine SQL bind `tenant_id` (line 257, 289). Code review verified; PG-real test deferred to T6 per PR body |
| 5 | Optional `source` / `kind` filters | ✅ | `CAST(:source AS text) IS NULL OR source = :source` pattern in both queries |
| 6 | `_coerce_uuid` accepts UUID + hex strings (driver portability) | ✅ | `test_coerce_uuid_*` |
| 7 | CI green | ❌ | **2 CI failures, 1 real**: (a) SonarCloud Quality Gate FAIL on "C Reliability Rating on New Code" — driven by B1 below; (b) `Python (ruff + mypy + pytest)`, `Build image`, `Semgrep SAST` all FAIL on docker registry rate-limit / testcontainers infrastructure flakes (same pattern as PR 296/298 — pre-existing infra, not code). Local gates all green (373 passed). |

---

### Project-board alignment

| Issue | Repo | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #261 | evoila/meho | `Closes #261` | ✅ same-repo | ✅ full modulo B1/B2/M1 + CI gate | Close on merge once findings + Sonar QG clear |
| #225 (initiative) | evoila/meho | parent of #261 | ❌ remains open | partial — T1 (#258), T2 (#259), T3 (#260) merged; T4 (#261, this PR); T5-T6 (#262-#263) pending on stacked branches | remains open |
| #221 (goal) | evoila/meho | parent of #225 | ❌ remains open | partial | remains open |

PR is correctly stacked on T3's merge commit (`4ec105d` per CR's commit range).

---

### Predicted vs. actual diff

Predicted from issue #261: `retrieval/retriever.py` (new), tests, docs. 3 files predicted, 3 files in PR — 1:1 match.

- ✅ Expected and correct (3 files): `backend/src/meho_backplane/retrieval/retriever.py` (+ ~430 lines), `backend/tests/test_retrieval_retriever.py` (+ ~280 lines), `docs/codebase/backend.md` (+1)
- 🚫 No undisclosed scope
- 🕳️ No missing files

The PR body explicitly defers the PG-real SQL coverage to T6's integration suite (`tests/integration/test_retrieval_e2e.py`), citing the cross-loop asyncpg cleanup pattern that needs T6's async-fixture-managed event loop. Reasonable scope-split.

---

### Integration verification (in review worktree on `b260196`)

- [x] `ruff check src/ tests/` — clean
- [x] `ruff format --check src/ tests/` — 82 files formatted
- [x] `mypy src/ --ignore-missing-imports` — 42 source files (+1 vs pre-T4: `retriever.py`), no issues
- [x] `pytest tests/ -q --deselect …test_metrics_endpoint… --ignore=tests/integration` — **373 passed**, 1 skipped (slow T2 real-model), 1 deselected (macOS flake)
- [❌ CI real] `SonarCloud Code Analysis` — FAILURE, Quality Gate "C Reliability Rating on New Code" caused by **B1** (S1244 float-eq, line 256). Fixing B1 clears the gate.
- [❌ CI infra] `Python (ruff + mypy + pytest)` — FAILURE on `tests/integration/test_audit_rows_correctly_scoped_per_tenant` docker create 404 (testcontainers); same pre-existing flake hitting every recent PR.
- [❌ CI infra] `Build image` — Docker Hub `429 Too Many Requests` on python base image (same as PR 298).
- [❌ CI infra] `Semgrep SAST` — Docker Hub rate-limit during `Initialize containers`; the actual Semgrep scan never ran.
- [x] CI green: `Go`, `Helm`, `DCO`, `TruffleHog`
- ⚠️ PG-side SQL exercised — not run (deferred to T6's integration suite per PR body, citing the asyncpg cleanup-loop issue)

---

### Findings (inline — each is a merge blocker per team policy)

For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 **B1** at `tests/test_retrieval_retriever.py:256` (this skill's inline) — `assert e.fused_score == 0.0` triggers Sonar `S1244` (MAJOR/BUG) and is the sole driver of the SonarCloud Quality Gate FAIL on "C Reliability Rating on New Code". Required change: `pytest.approx(0.0)` — pattern already used at lines 195, 213, 230, etc.
- 🛑 **B2 / CR-2** at `tests/test_retrieval_retriever.py:67` (already inline from [CR comment 3228911954](https://github.com/evoila/meho/pull/299#discussion_r3228911954) + Sonar S5655 CRITICAL/CODE_SMELL at line 191) — `_row(doc_id, score: float)` is typed `float` but tests pass `None` for the no-bm25-score path. Required change: `score: float | None`. Not duplicated by this skill — CR's inline carries the action item.
- ⚠️ **M1 / CR-1** at `src/meho_backplane/retrieval/retriever.py:147-154` (already inline from [CR comment 3228911950](https://github.com/evoila/meho/pull/299#discussion_r3228911950)) — `retrieve()` doesn't validate `limit` before passing to `[:limit]` slicing. Negative → no-op truncation; zero → empty result via slicing semantics (works by accident). Required change: explicit guard — `raise ValueError` on negative, early-return `[]` on zero. Real defense-in-depth concern; the helper is framed as "the shared read path G4 and G5 both consume" so production-correctness matters. Not duplicated by this skill.

This skill's own adversarial pass found one additional finding (B1, the Sonar QG driver) on top of CR's two inlines.

---

### Adjacent improvements (non-blocking, not posted inline)

- **PG-side SQL coverage deferred to T6.** The PR body's rationale is sound (asyncpg cleanup loop), but the deferral does mean the load-bearing `to_tsvector('english', ...)` literal, the `documents_body_fts_idx` GIN index, the IVFFlat probe-list parameters, and the `<=>` cosine-distance shape don't get a real-world smoke test on this PR. T6's integration suite (#263) needs to exercise these.
- **Source-language hardcode (`'english'`).** The BM25 SQL pins the tsquery language to `'english'`. Tenants with non-English corpora would silently get degraded recall. v0.2-spec scope per the Initiative, but worth a follow-up issue under G4 / G5 once non-English consumers actually surface.

---

### AI reviewer (CodeRabbit) cross-check

- **Findings adopted (CR's inlines stay; not duplicated):** 2 (CR-1 = M1 limit guard; CR-2 = B2 `_row` type)
- **Deferred to Adjacent improvements:** 0
- **Disputed:** 0

This skill added one inline (B1, the Sonar QG driver) that CR didn't surface — same pattern of CR missing the SonarCloud-specific float-equality lint that I observed on PR 296.

---

### Docs

- [x] `docs/codebase/backend.md` — one-line addition for the new retriever module. Status: `updated`.

---

### What worked well

- **RRF fusion math isolated as a pure function and exhaustively tested.** 11 unit tests cover the load-bearing behaviour (empty inputs, BM25-only, cosine-only, intersection outranking singletons, limit truncation, None-score gracefulness) without needing pgvector or a real embedding model.
- **`Why RRF and not weighted-sum` rationale documented at top of module + in PR body.** Saves the next reviewer asking "why didn't we just normalise + weighted-sum?".
- **Tenant scoping is required-parameter, not inferred from contextvars.** Two distinct SQL statements both bind `tenant_id`; the call site stays auditable. Same pattern T4 (#249) established for resource handlers, applied consistently here.
- **`_coerce_uuid` driver-portability helper.** Both asyncpg (returns UUID) and SQLite (returns str) work through the same fusion path without per-driver branching elsewhere in the module.
- **Explicit `CANDIDATE_LIMIT = 50` constant with tuning-rationale comment.** The "tune upward when corpora grow past ~10k documents per tenant" guidance is exactly the operational knowledge dashboards-and-runbooks will want.

**Comment by ikaric** [comment]
## Review feedback addressed — pushed in `476689d`

Force-push of T4 (with-lease). All findings resolved:

### Blockers (Damir)

- **B1 — `python:S1244` exact float equality.** `tests/test_retrieval_retriever.py:263` (`e.fused_score == 0.0`) switched to `pytest.approx(0.0)`. Matches the style every other float assertion in the file uses (lines 195, 213, 230 etc.) and clears the SonarCloud Quality Gate FAIL on "C Reliability Rating on New Code".

### CodeRabbit majors

- **CR-1 — `retrieve()` limit guard.** Added explicit validation: `limit < 0` raises `ValueError("limit must be >= 0; got <n>")`; `limit == 0` short-circuits to `[]` without embedding or SQL. Same guards added inside `_rrf_fuse` since unit tests exercise the pure helper directly. Two new tests pin both branches:
  - `test_rrf_fuse_rejects_negative_limit` — `ValueError` on negative.
  - `test_rrf_fuse_zero_limit_returns_empty_without_sorting` — empty list on zero.
- **CR-2 — `_row` type signature.** Updated `def _row(doc_id: uuid.UUID, score: float | None) -> Any` to match the `None`-score test case at line ~191 (and the real PG shape — `ts_rank_cd` returns NULL against degenerate tsqueries). Docstring expanded to call out why `None` is in scope.

### Pre-existing CI failures (not introduced here)

- `Python (ruff + mypy + pytest)` testcontainers docker-create 404 — same flake hitting recent PRs, infra-side.
- `Build image` Docker Hub `429` — pull rate-limit, infra-side.
- `Semgrep SAST` rate-limit on the initialize-containers step — infra-side.

### Validation

- `ruff check .` clean (86 files).
- `ruff format --check .` clean.
- `mypy --strict src/` clean across 42 source files.
- `pytest tests/` — 372 passed (+2 from the new limit-guard tests, vs. 370 before), 1 skipped (slow), 1 pre-existing macOS-only `test_observability` failure unchanged.
- T5/T6 cascade-rebased onto the amended T4; whole local stack stays green.
